sdb/command.py

```diff
@@ -225,14 +234,104 @@ def __init_subclass__(cls, **kwargs):
 
     def _call(self,
               objs: Iterable[drgn.Object]) -> Optional[Iterable[drgn.Object]]:
-        # pylint: disable=missing-docstring
-        raise NotImplementedError
+        """
+        Implemented by the subclass.
+        """
+        raise NotImplementedError()
+
+    def __invalid_memory_objects_check(self, objs: Iterable[drgn.Object],
+                                       fatal: bool) -> Iterable[drgn.Object]:
+        """
+        A filter method for objects passed through the pipeline that
+        are backed by invalid memory. When `fatal` is set to True
+        we raise an error which will stop this control flow when
+        such objects are encountered. If `fatal` is False we just
+        print the error and go on.
+        """
+        for obj in objs:
+            try:
+                obj.read_()
+            except drgn.FaultError as err:
+                if obj.address_ is None:
+                    #
+                    # This is possible when the object was created `echo`.
+                    #
+                    err_msg = str(err)
+                else:
+                    err_msg = f"addresss {hex(obj.address_of_().value_())}"
+                err = CommandError(self.name,
+                                   f"invalid memory access: {err_msg}")
+                if fatal:
+                    raise err
+                print(err.text)
+                continue
+            yield obj
 
     def call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
         # pylint: disable=missing-docstring
-        result = self._call(objs)
-        if result is not None:
-            yield from result
+        #
+        # Even though we have __invalid_memory_objects_check() to
+        # ensure that the objects returned are valid, we still
+        # need to account for invalid accesses happening while
+        # the command is running.
+        #
+        try:
+            result = self._call(objs)
+            if result is not None:
+                #
+                # The whole point of the SingleInputCommands are that
+                # they don't stop executing in the first encounter of
+                # a bad dereference. That's why we check here whether
+                # the command that we are running is a subclass of
+                # SingleInputCommand and we set the `fatal` flag
+                # accordinly.
+                #
+                yield from self.__invalid_memory_objects_check(
+                    result, not issubclass(self.__class__, SingleInputCommand))
+        except drgn.FaultError as err:
+            raise CommandError(self.name, f"invalid memory access: {str(err)}")
+
+
+class SingleInputCommand(Command):
+    """
+    Commands that are expected to recover from bad memory accesses
+    implicitly by the sdb infrastructure and move on to the next
+    object should subclass this class.
```

Comment:
I think the bad memory access recovery is nice but maybe we can put the principle of processing each object independently up front.  I could imagine wanting to recover from additional errors in the future (e.g. ArrayBoundsError, once we do that):
`Commands that would like to process each input object independently (without saving state between objects) can subclass from this class.  If a FaultError exception is thrown while processing one object, processing will continue with the next object.`